util/decimal: decrease allowed shift length

madelynnblue opened the pull request:
Improve decimal testing to include a 5s timeout. Also use subtests for
better test messages.

Fixes #10928

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10934)
<!-- Reviewable:end -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/util/decimal/decimal.go, line 337 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10934#-KXA1MZc1Hj95dpAkngi:-KXA1MZc1Hj95dpAkngj:b-x7zotu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b56619ae652fdb393a16a2ab5d3e7e967f00ba2f/pkg/util/decimal/decimal.go#L337)):*
> ```Go
> 	// this it becomes very slow. However since the implementation for that lies
> 	// in the inf package, we instead check the length in this function. The
> 	// tests in TestDecimalLogN that using strE (a very high-precision E value)
> ```

I think a word is missing here. Perhaps "the tests ... that *are* using strE" ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10934#-:-KXA1ipFCm7kzoiAibm-:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 2 files at r2.
Review status: 3 of 4 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/util/decimal/decimal_test.go, line 87 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10934#-KXARFlHjUl99mT9Zowi:-KXARFlHjUl99mT9Zowj:b6y0opy) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd52170cfdb2b95c2b783d794da6944923f8455a/pkg/util/decimal/decimal_test.go#L87)):*
> ```Go
> ) {
> 	for _, tc := range tests {
> 		t.Run(fmt.Sprintf("%s = %s", tc.input, tc.expected), func(t *testing.T) {
> ```

it's a good idea to avoid spaces in these names. run this test with `-test.v` to see why.

---

*[pkg/util/decimal/decimal_test.go, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10934#-KXARYCEuRTup2mnmZsK:-KXARYCEuRTup2mnmZsL:b-4sb2un) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd52170cfdb2b95c2b783d794da6944923f8455a/pkg/util/decimal/decimal_test.go#L105)):*
> ```Go
> 				t.Logf("execute duration: %s", timeutil.Since(start))
> 			case <-time.After(testFuncTimeout):
> 				t.Fatalf("timedout after %s", testFuncTimeout)
> ```

won't this leak goroutines? whoa, is the parser package no using leaktest?

---

*[pkg/util/decimal/decimal_test.go, line 111 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10934#-KXARH9V5z161fyJuj4I:-KXARH9V5z161fyJuj4J:b-kbnyvh) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd52170cfdb2b95c2b783d794da6944923f8455a/pkg/util/decimal/decimal_test.go#L111)):*
> ```Go
> 					t.Errorf("expected error %s, got %s", tc.expected, err)
> 				}
> 				return
> ```

not new but this seems to do the wrong thing if an error is expected and not returned.

---

*[pkg/util/decimal/decimal_test.go, line 157 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10934#-KXAQzhiYZTTzLibmLXX:-KXAQzhiYZTTzLibmLXY:b6y0opy) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd52170cfdb2b95c2b783d794da6944923f8455a/pkg/util/decimal/decimal_test.go#L157)):*
> ```Go
> ) {
> 	for _, tc := range tests {
> 		t.Run(fmt.Sprintf("%s, %s = %s", tc.input1, tc.input2, tc.expected), func(t *testing.T) {
> ```

it's a good idea to avoid spaces in these names. run this test with `-test.v` to see why.

---

*[pkg/util/decimal/decimal_test.go, line 161 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10934#-KXAQw5010ONlE39qbrn:-KXAQw5010ONlE39qbro:bqkgyvz) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd52170cfdb2b95c2b783d794da6944923f8455a/pkg/util/decimal/decimal_test.go#L161)):*
> ```Go
> 			if _, ok := x.SetString(tc.input1); !ok {
> 				t.Errorf("could not set decimal: %s", tc.input1)
> 				return
> ```

this is equivalent to `Fatal`ing on the previous line

---

*[pkg/util/decimal/decimal_test.go, line 165 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10934#-KXAQzPd72s2tDMlWeb4:-KXAQzPd72s2tDMlWeb5:bqkgyvz) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd52170cfdb2b95c2b783d794da6944923f8455a/pkg/util/decimal/decimal_test.go#L165)):*
> ```Go
> 			if _, ok := y.SetString(tc.input2); !ok {
> 				t.Errorf("could not set decimal: %s", tc.input2)
> 				return
> ```

this is equivalent to `Fatal`ing on the previous line

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10934)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 2 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/util/decimal/decimal.go, line 337 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10934#-KXA1MZc1Hj95dpAkngi:-KXCF4HPwELMiczsER-s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b56619ae652fdb393a16a2ab5d3e7e967f00ba2f/pkg/util/decimal/decimal.go#L337)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I think a word is missing here. Perhaps "the tests ... that *are* using strE" ?
</details>
Done.

---

*[pkg/util/decimal/decimal_test.go, line 87 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10934#-KXARFlHjUl99mT9Zowi:-KXCF5xpmY4pTQ8P4V85:bouv5at) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd52170cfdb2b95c2b783d794da6944923f8455a/pkg/util/decimal/decimal_test.go#L87)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> it's a good idea to avoid spaces in these names. run this test with `-test.v` to see why.
</details>
I prefer the underscores myself, but I've changed these.

---

*[pkg/util/decimal/decimal_test.go, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10934#-KXARYCEuRTup2mnmZsK:-KXCFFYr1ANeZu-PEY_x:b-dp6ovl) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd52170cfdb2b95c2b783d794da6944923f8455a/pkg/util/decimal/decimal_test.go#L105)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> won't this leak goroutines? whoa, is the parser package no using leaktest?
</details>
Changed to use a timer and Stop. Should prevent leaks.

---

*[pkg/util/decimal/decimal_test.go, line 111 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10934#-KXARH9V5z161fyJuj4I:-KXCahR8FbyrLHcbTKCb:bgly6re) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd52170cfdb2b95c2b783d794da6944923f8455a/pkg/util/decimal/decimal_test.go#L111)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> not new but this seems to do the wrong thing if an error is expected and not returned.
</details>
I think it's ok. If an error is expected, it will fail to set and compare during exp.SetString and exp.Cmp later down since tc.expected is an error string, not a valid decimal.

---

*[pkg/util/decimal/decimal_test.go, line 161 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10934#-KXAQw5010ONlE39qbrn:-KXCb5apirn3hcdTvrvv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd52170cfdb2b95c2b783d794da6944923f8455a/pkg/util/decimal/decimal_test.go#L161)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is equivalent to `Fatal`ing on the previous line
</details>
Done.

---

*[pkg/util/decimal/decimal_test.go, line 165 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10934#-KXAQzPd72s2tDMlWeb4:-KXCb5pZS3QXOkvdPnpg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd52170cfdb2b95c2b783d794da6944923f8455a/pkg/util/decimal/decimal_test.go#L165)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is equivalent to `Fatal`ing on the previous line
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10934)*
<!-- Sent from Reviewable.io -->

mjibson commented:
PTAL

knz commented:
Reviewed 1 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/util/decimal/decimal_test.go, line 161 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10934#-KXAQw5010ONlE39qbrn:-KXR3vyGgex8vxYTxatW:b2ajdbo) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd52170cfdb2b95c2b783d794da6944923f8455a/pkg/util/decimal/decimal_test.go#L161)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> Done.
</details>
I may be missing something; you have made two such substitutions below this point but this one (on line 162-163) and the following seem to remain.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10934)*
<!-- Sent from Reviewable.io -->

knz commented:
Also please check (possibly manually) that this change properly addresses the latest repro `SELECT pow(2.306257620454719e+12::decimal, 49.18687811476825::decimal)`.

knz commented:
Another repro: `SELECT POW(791018.4038517432::decimal, 155.94813858582634::decimal)`

madelynnblue committed:
util/decimal: decrease allowed shift length

Improve decimal testing to include a 5s timeout. Also use subtests for
better test messages.

Test timeouts need to be much longer during race tests.

Fixes #10928

mjibson commented:
I've also added those other two failing RSG tests. They pass here.

---

Review status: 3 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/util/decimal/decimal_test.go, line 161 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10934#-KXAQw5010ONlE39qbrn:-KXdipv7F4LKLONwxdAY:b-v59cgu) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd52170cfdb2b95c2b783d794da6944923f8455a/pkg/util/decimal/decimal_test.go#L161)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I may be missing something; you have made two such substitutions below this point but this one (on line 162-163) and the following seem to remain.
</blockquote></details>

Yeah not sure how that didn't get picked up. I thought I already had this. Added.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10934#-:-KXdj47zNklEtQhunsiC:bqg42lo)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10934#-:-KXdl5XZZFrzJLUBW1nU:bnfp4nl)*
<!-- Sent from Reviewable.io -->
